### ISSUE TITLE
Vega updates for streaming data do not work on multi-layer facet plots

### ISSUE THREAD

**Issue by ameyabp** [vega/vega-lite#7597]
I am creating a multi-layer faceted bar chart for visualizing progressive data. I visualize a configurable sliding window of past data samples, where a data sample is identified by the sample number assigned to it. I have a nominal attribute which I map to the facet column, the independent variable to the Y axis, and the sample number to the X axis. I have 2 layers in my plot - one layer visualizes the bar marks for the basic bar chart, and the other layer visualizes rect marks for the confidence intervals. I fix the domain of the X axis to the width of the sliding window in the spec, and accordingly update the sample number of all the samples in the vega view to maintain the correct number of latest bars in the window. I use vega view API to update the data. However, the updates to the sample number only reflect in the layer having the bar marks. The layer having the rect marks do not see the updated sample numbers and end up overplotting on top of each other.

I am using vega 5.20.2, vega-lite 5.1.0 and vega-embed 6.18.2

Here is an example JSON spec. It is a snapshot of one of the sliding windows in the visualization. This spec works fine in the online vega editor.
```
{
  "$schema": "https://vega.github.io/schema/vega-lite/v5.json",
  "data": {
    "name": "table",
    "values": 
[
    {
        "dataset": "A",
        "sampleNumber": -9,
        "sampleSize": 0.9296,
        "mean": 14.322873707746728,
        "ci_low": 13.700866237359504,
        "ci_hi": 15.148570561053338,
        "ci_low_100": 13.237363334438482,
        "ci_hi_100": 15.90399630783704
    },
    {
        "dataset": "B",
        "sampleNumber": -9,
        "sampleSize": 0.9296,
        "mean": 10.945206872197998,
        "ci_low": 10.122924403405602,
        "ci_hi": 13.812928943524991,
        "ci_low_100": 9.889248258834565,
        "ci_hi_100": 14.262169260177647
    },
    {
        "dataset": "A",
        "sampleNumber": -8,
        "sampleSize": 0.9379,
        "mean": 14.30787090703482,
        "ci_low": 13.703858959955253,
        "ci_hi": 15.14613284904854,
        "ci_low_100": 13.245509957412517,
        "ci_hi_100": 15.86676400499494
    },
    {
        "dataset": "B",
        "sampleNumber": -8,
        "sampleSize": 0.9379,
        "mean": 10.935174387291605,
        "ci_low": 10.11882504114157,
        "ci_hi": 13.843101170767788,
        "ci_low_100": 9.874635810870078,
        "ci_hi_100": 15.363364648608545
    },
    {
        "dataset": "A",
        "sampleNumber": -7,
        "sampleSize": 0.9462,
        "mean": 14.310685843574763,
        "ci_low": 13.726387342494226,
        "ci_hi": 15.170628526813326,
        "ci_low_100": 13.272445471065883,
        "ci_hi_100": 15.790215005926687
    },
    {
        "dataset": "B",
        "sampleNumber": -7,
        "sampleSize": 0.9462,
        "mean": 10.916570901665734,
        "ci_low": 10.109749464178375,
        "ci_hi": 14.189713029226569,
        "ci_low_100": 9.871784248716377,
        "ci_hi_100": 14.589886878954102
    },
    {
        "dataset": "A",
        "sampleNumber": -6,
        "sampleSize": 0.9545,
        "mean": 14.323332290885109,
        "ci_low": 13.719947711132708,
        "ci_hi": 15.125890089309166,
        "ci_low_100": 13.283348349403106,
        "ci_hi_100": 16.012663082133038
    },
    {
        "dataset": "B",
        "sampleNumber": -6,
        "sampleSize": 0.9545,
        "mean": 10.908072417042595,
        "ci_low": 10.121614812557679,
        "ci_hi": 13.81533903543392,
        "ci_low_100": 9.864797544078293,
        "ci_hi_100": 14.456381679950947
    },
    {
        "dataset": "A",
        "sampleNumber": -5,
        "sampleSize": 0.9628,
        "mean": 14.297534976411297,
        "ci_low": 13.695521213054137,
        "ci_hi": 15.072229178581836,
        "ci_low_100": 13.309213526707065,
        "ci_hi_100": 15.77141926863209
    },
    {
        "dataset": "B",
        "sampleNumber": -5,
        "sampleSize": 0.9628,
        "mean": 10.87500032777771,
        "ci_low": 10.082047736630162,
        "ci_hi": 13.685285977242295,
        "ci_low_100": 9.840171520497993,
        "ci_hi_100": 14.126296311778374
    },
    {
        "dataset": "A",
        "sampleNumber": -4,
        "sampleSize": 0.9711,
        "mean": 14.28266021845597,
        "ci_low": 13.67805032526131,
        "ci_hi": 15.066175912262642,
        "ci_low_100": 13.19583325657908,
        "ci_hi_100": 15.878189194541672
    },
    {
        "dataset": "B",
        "sampleNumber": -4,
        "sampleSize": 0.9711,
        "mean": 10.858399756871265,
        "ci_low": 10.068163379263273,
        "ci_hi": 13.617237056098515,
        "ci_low_100": 9.849103808112668,
        "ci_hi_100": 14.803913103779157
    },
    {
        "dataset": "A",
        "sampleNumber": -3,
        "sampleSize": 0.9794,
        "mean": 14.298044440811239,
        "ci_low": 13.696436376251839,
        "ci_hi": 15.095896700668598,
        "ci_low_100": 13.303121498425181,
        "ci_hi_100": 16.021256960637732
    },
    {
        "dataset": "B",
        "sampleNumber": -3,
        "sampleSize": 0.9794,
        "mean": 10.83540581160387,
        "ci_low": 10.059430872813714,
        "ci_hi": 13.975512829826501,
        "ci_low_100": 9.827721751178808,
        "ci_hi_100": 15.461338041143916
    },
    {
        "dataset": "A",
        "sampleNumber": -2,
        "sampleSize": 0.9877,
        "mean": 14.282700859921139,
        "ci_low": 13.696113817563054,
        "ci_hi": 15.093919005482926,
        "ci_low_100": 13.2789846647872,
        "ci_hi_100": 15.7251098055895
    },
    {
        "dataset": "B",
        "sampleNumber": -2,
        "sampleSize": 0.9877,
        "mean": 10.820613484921324,
        "ci_low": 10.052656631345476,
        "ci_hi": 13.474311623570722,
        "ci_low_100": 9.815836073517188,
        "ci_hi_100": 14.410457900751993
    },
    {
        "dataset": "A",
        "sampleNumber": -1,
        "sampleSize": 0.996,
        "mean": 14.26943470625928,
        "ci_low": 13.694180768539294,
        "ci_hi": 15.075789649695357,
        "ci_low_100": 13.24913040932822,
        "ci_hi_100": 15.665325424735409
    },
    {
        "dataset": "B",
        "sampleNumber": -1,
        "sampleSize": 0.996,
        "mean": 10.901418705687666,
        "ci_low": 10.139244365554744,
        "ci_hi": 13.727900578937815,
        "ci_low_100": 9.882231470403074,
        "ci_hi_100": 14.478111692556562
    },
    {
        "dataset": "A",
        "sampleNumber": 0,
        "sampleSize": 1.0043,
        "mean": 14.252864197723053,
        "ci_low": 13.663734009087184,
        "ci_hi": 15.043612257158479,
        "ci_low_100": 13.225317699449235,
        "ci_hi_100": 15.805971277881081
    },
    {
        "dataset": "B",
        "sampleNumber": 0,
        "sampleSize": 1.0043,
        "mean": 10.889926410809611,
        "ci_low": 10.122497248812383,
        "ci_hi": 13.668696798663433,
        "ci_low_100": 9.867534215428849,
        "ci_hi_100": 14.08234762943323
    }
]



  },
  "facet": {
    "column": {
      "field": "dataset",
      "header": {"title": null, "labelFontSize": 15, "labelOrient": "bottom"}
    }
  },
  "spec": {
    "width": 200,
    "height": 200,
    "layer": [
      {
        "mark": {"type": "bar", "align": "center", "width": 10, "opacity": 0.7},
        "encoding": {
          "x": {
            "field": "sampleNumber",
            "type": "quantitative",
            "title": "",
            "axis": null,
            "scale": {"domain": [-9.5, 0.5]}
          },
          "y": {
            "field": "mean",
            "type": "quantitative",
            "title": "Salary (K€)",
            "axis": {
              "titleFontSize": 15,
              "labelFontSize": 15,
              "tickMinStep": 10
            },
            "scale": {"domain": [0, 34]}
          },
          "color": {
            "field": "sampleNumber",
            "type": "quantitative",
            "scale": {"scheme": "blues"},
            "legend": null
          }
        }
      },
      {
        "mark": {
          "type": "rect",
          "width": 5,
          "align": "center",
          "color": "red",
          "opacity": 0.7
        },
        "encoding": {
          "x": {
            "field": "sampleNumber",
            "type": "quantitative",
            "title": null,
            "axis": null,
            "scale": {"domain": [-9.5, 0.5]}
          },
          "y": {"field": "ci_low", "type": "quantitative", "title": ""},
          "y2": {"field": "ci_hi", "type": "quantitative"}
        }
      }
    ]
  },
  "padding": {"top": 20, "right": 20, "bottom": 20, "left": 20}
}
```

This is how it looks in the editor - the expected visualization.
![vegaEditor](https://user-images.githubusercontent.com/43589244/125806593-a6c65292-45cc-4071-8136-d5c92b9dee10.png)

Following is the code that I use to update the data elements using vega view API. Every new data sample added to the view is assigned a sample number of 0 and then as newer updates arrive, the sample number for the old existing updates is decreased, till they are removed from the view completely.
```
function updateVega(vegaView, values, barNames, slidingWindowSize) {
    let entry = []
    for (const i in barNames) {
        values[i]["sampleNumber"] = 0;
        entry.push({...values[i], dataset: barNames[i]});
    }

    let changeSet = vega.changeset()
    changeSet = changeSet.remove(d => {
        d.sampleNumber -= 1;
        return d.sampleNumber < 1 - slidingWindowSize;
    });

    changeSet = changeSet.insert(entry)
    vegaView.change("table", changeSet).runAsync().then();
}
```

This is how it looks with the above vega update code - notice the overplotting of the CI bars. The bar marks are rendered correctly, but the CI rects are not, even though they are both encoded to the same attribute 'sampleNumber'.
![vegaViewAPI](https://user-images.githubusercontent.com/43589244/125805911-7100f036-702d-433b-8699-77de0937c33b.png)

Basically, the underlying data is the same, but the changes are reflected in only one of the 2 layers. Additionally, it turns out, the changes do not work even if I retain only the CI bars layer and remove the bar mark layer.

Please let me know if any additional information is needed. I would be happy to help

**Comment by domoritz** [comment]
Thank you for the bug report. Could you look at the generated Vega spec and see whether you can manually fix the issue there?

**Comment by ameyabp** [comment]
Sorry, I did not get you there when you say 'manually fix the issue'. Any snapshot of the continuously updating vega spec does not show the data updates, it shows the original spec with which the visualization was setup. So I cannot manually fix the issue by looking at the vega spec (unless I am missing something that I clearly don't understand). The spec that I provided in the bug report is just an example of how it should have looked. The fact that the bars show up correctly also proves that the updated data in the spec is as expected. Its just that for some reason, the CI rect marks do not manifest that mapping.

**Comment by ameyabp** [comment]
Update - we were able to work around this issue by specifying the domain using 'domainMin' and 'domainMax', instead of specifying it as an array of 2 elements. 

So instead of this
`
"scale": {"domain": [-9.5, 0.5]}
`
we did this
`
"scale": {"domainMin": -9.5, "domainMax": 0.5}
`
and it worked. It looks like a problem of ambiguous interpretation of the array styled specification of the domain

**Comment by domoritz** [comment]
Nice. Can you boil this down to a minimal example? Maybe it's a bug in Vega.

**Comment by ameyabp** [comment]
I was trying to create a reproducible minimal example of this bug, however I am unable to reproduce it anymore. The bug does not show up even in the original code.
